### PR TITLE
Toggle dark/light mode manually

### LINKED ISSUE #125: Toggle dark mode manually
The current `tailwind-rn` only supports dark mode when it set on the system OS. It would be nice to have support for passing color scheme to the TailwindProvider to enable manually toggling color scheme. Any plans to add this or should I make a PR?

### PR DESCRIPTION
**Pull request by catalinmiron**
Closes #125 

@vadimdemedes I couldn't find a better name for the prop and I stick with `theme`, maybe you have a suggestion :)

### CODE REVIEW

File: source/tailwind-provider.tsx [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
 
 interface Props {
 	utilities: Utilities;
+	theme?: ColorSchemeName;
 }
 
-const TailwindProvider: React.FC<Props> = ({utilities, children}) => {
-	const colorScheme = useColorScheme() ?? 'light';
+const TailwindProvider: React.FC<Props> = ({utilities, theme, children}) => {
+	const colorScheme = theme ?? useColorScheme() ?? 'light';
```
Comment (HeleneGrini) [inline source/tailwind-provider.tsx:19]: shouldn't conditionally call hooks. Add this logic in the  useMemo instead  

```
	const colorScheme = useColorScheme() ?? 'light';
        ...
       	const tailwind = React.useMemo(() => {
		return create(utilities, {
			colorScheme: theme ?? colorScheme,
			width,
			height,
			reduceMotion: Boolean(reduceMotion),
			orientation
		});
	}, [utilities, colorScheme, width, height, reduceMotion, orientation, theme]);

Comment (catalinmiron) [reply source/tailwind-provider.tsx:19]: Hi @HeleneGrini, thanks for reviewing it. What is the difference? 
In case theme or colorScheme has changed, tailwind will change as well. Maybe I'm missing something, but could you please extend your thoughts a little? Thanks!

Comment (HeleneGrini) [reply source/tailwind-provider.tsx:19]: Thanks for making the PR! 😄 

It's one of the rules of hooks -- they shouldn't be called in loops, conditions, or  functions (other than hooks or komponents) https://reactjs.org/docs/hooks-rules.html 

It can result in the order/number of hook being different on different renders, which will result in an error as React stores the value of hooks in an ordered manner. If the order changes, the hooks will not me matched to their appropriate value.   



An example here would be that you have a state in your app storing the `theme`. Let's say your state is `undefined` unless the user has manually set a theme in you app. This will then default to the system colorScheme in the TailwindProvider. `useColorScheme` gets called, then `useMemo` gets called. Later your user sets the theme in your app, the props to the TailwindProvider change causing a rerender. On this render theme props is set and `useColorScheme` is _not_ called. `useMemo` gets called as the first hook, and will the be matched to first stored hook-value in react, but that was expected to be `useColorScheme` as it was the previous render. Booom, error 

I would suggest having a look at this explanation https://reactjs.org/docs/hooks-rules.html#explanation

Comment (catalinmiron) [reply source/tailwind-provider.tsx:19]: Hi @HeleneGrini, interesting topic here. Maybe I misunderstood everything, but I am changing a variable at runtime. Regarding your example, anytime either `colorScheme` and changed (either via `useColorScheme` or `theme`), tailwind with be changed as well. 

Let's say:
```
useColorScheme() returns 'light';
theme = undefined;

`create` is called and tailwind will return light theme styles;

user change theme to `light` => `create` will not be called, because colorScheme is equal to light, therefore there's no need to update it.
user change theme to 'dark' => `create` is called, because `colorScheme` has changed, prev light, current dark;
```

or 

```
useColorScheme() returns undefined;
theme = undefined;
colorScheme will be `light` (the fallback)

`create` is called and tailwind will return light theme styles;

and now follows the flow presented from above.
```

Maybe I don't understand this concept enough and I can be completely wrong.

`colorScheme` its a variable + a dependency of useMemo which can take a value of `theme` or `useColorScheme` or `light`, we don't call a hook conditionally here.

Comment (HeleneGrini) [reply source/tailwind-provider.tsx:19]: The issue is not with the tailwind variable, but rather the colorScheme variabel 

```
const colorScheme = theme ?? useColorScheme() ?? 'light';
```
 if `theme` from props is `undefined`  `useColorScheme` is called, if `theme` from props is set `useColorScheme`  is _not_ called hence you are conditionally calling the hook. 
 
 The value of `theme` from props may change from `undefined` to `'light' | 'dark'` or vice versa, which would result in an error

Comment (catalinmiron) [reply source/tailwind-provider.tsx:19]: Oh dayum, thanks for the clarification. I am going tonpush an update. 🤗

Comment (catalinmiron) [reply source/tailwind-provider.tsx]: Done, please take another look. Thanks!

File: readme.md [hunk 2]
```diff
@@ -183,7 +183,24 @@ import {TailwindProvider} from 'tailwind-rn';
 import utilities from './tailwind.json';
 
 const App = () => (
-	<TailwindProvider utilities={utilities}>
+	<TailwindProvider utilities={utilities} theme="dark">
```
Comment (vadimdemedes) [inline readme.md:186]: This example isn't related to the introduction of the new prop, could you undo the change here?

File: readme.md [hunk 3]
```diff
@@ -183,7 +183,24 @@ import {TailwindProvider} from 'tailwind-rn';
...
+		<MyComponent />
+	</TailwindProvider>
+);
+```
+
+#### theme
```
Comment (vadimdemedes) [inline readme.md:192]: Could you rename the prop to `colorScheme` to match the naming in the codebase?

File: readme.md [hunk 4]
```diff
@@ -183,7 +183,24 @@ import {TailwindProvider} from 'tailwind-rn';
...
+
+Type: `string | undefined | null`
```
Comment (vadimdemedes) [inline readme.md:194]:
```suggestion
Type: `string`
```

File: readme.md [hunk 5]
```diff
@@ -183,7 +183,24 @@ import {TailwindProvider} from 'tailwind-rn';
...
+
+Manually set theme mode: `light` or `dark`. This prop will override the native color scheme and its an optional prop.
```
Comment (vadimdemedes) [inline readme.md:196]:
```suggestion
Override the default color scheme. Possible values are `light` or `dark`.
```

File: source/tailwind-provider.tsx [hunk 6]
```diff
@@ -1,19 +1,21 @@
...
 import {
-	useDimensions,
 	useAccessibilityInfo,
-	useDeviceOrientation
+	useDeviceOrientation,
+	useDimensions
 } from '@react-native-community/hooks';
-import TailwindContext from './tailwind-context';
+import * as React from 'react';
+import {useColorScheme, ColorSchemeName} from 'react-native';
```
Comment (vadimdemedes) [inline source/tailwind-provider.tsx:7]: Could you make sure there are only necessary changes here? Looks like the only change here is the new import for `ColorSchemeName`.

File: source/tailwind-provider.tsx [hunk 7]
```diff
@@ -1,19 +1,21 @@
...
+
 import create from './create';
+import TailwindContext from './tailwind-context';
 import {Utilities} from './types';
 
 interface Props {
 	utilities: Utilities;
+	theme?: ColorSchemeName;
```
Comment (vadimdemedes) [inline source/tailwind-provider.tsx:15]:
```suggestion
	colorScheme?: ColorSchemeName;
```

File: source/tailwind-provider.tsx [hunk 8]
```diff
@@ -1,19 +1,21 @@
...
 }
 
-const TailwindProvider: React.FC<Props> = ({utilities, children}) => {
+const TailwindProvider: React.FC<Props> = ({utilities, theme, children}) => {
```
Comment (vadimdemedes) [inline source/tailwind-provider.tsx:18]:
```suggestion
const TailwindProvider: React.FC<Props> = ({utilities, colorScheme: overrideColorScheme, children}) => {
```

File: source/tailwind-provider.tsx [hunk 9]
```diff
@@ -23,13 +25,13 @@ const TailwindProvider: React.FC<Props> = ({utilities, children}) => {
 
 	const tailwind = React.useMemo(() => {
 		return create(utilities, {
-			colorScheme,
+			colorScheme: theme ?? colorScheme,
```
Comment (vadimdemedes) [inline source/tailwind-provider.tsx:28]:
```suggestion
			colorScheme: overrideColorScheme ?? colorScheme,
```

File: source/tailwind-provider.tsx [hunk 10]
```diff
@@ -23,13 +25,13 @@ const TailwindProvider: React.FC<Props> = ({utilities, children}) => {
...
 			width,
 			height,
 			reduceMotion: Boolean(reduceMotion),
 			orientation
 		});
-	}, [utilities, colorScheme, width, height, reduceMotion, orientation]);
+	}, [utilities, colorScheme, width, height, reduceMotion, orientation, theme]);
```
Comment (vadimdemedes) [inline source/tailwind-provider.tsx:34]:
```suggestion
	}, [utilities, colorScheme, overrideColorScheme, width, height, reduceMotion, orientation]);
```

**Review by vadimdemedes — CHANGES_REQUESTED** [review]
Thanks for working on this!

**Comment by catalinmiron** [comment]
@vadimdemedes done. Please take another look. 

PS: Thanks for the suggestions!

File: source/tailwind-provider.tsx [hunk 11]
```diff
@@ -1,19 +1,25 @@
 import * as React from 'react';
-import {useColorScheme} from 'react-native';
+import {useColorScheme, ColorSchemeName} from 'react-native';
 import {
 	useDimensions,
 	useAccessibilityInfo,
 	useDeviceOrientation
 } from '@react-native-community/hooks';
+
```
Comment (vadimdemedes) [inline source/tailwind-provider.tsx:8]: Remove this empty line.

**Comment by vadimdemedes** [comment]
Great work, thank you!

**Comment by vadimdemedes** [comment]
Thanks for review help, @HeleneGrini 👍

**Comment by catalinmiron** [comment]
Thanks @vadimdemedes @HeleneGrini for the time! ❤️

**Comment by vadimdemedes** [comment]
Released in [v4.2.0](https://github.com/vadimdemedes/tailwind-rn/releases/tag/v4.2.0).
